### PR TITLE
Add EIP-5639: Delegation Registry

### PR DESCRIPTION
**Pull request by wwhchung**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Comment by abcoathup** [comment]
EIP numbers are assigned by EIP editors and are generally the PR number.
https://eips.ethereum.org/EIPS/eip-1#eip-editor-responsibilities

**Comment by wwhchung** [comment]
> EIP numbers are assigned by EIP editors and are generally the PR number. https://eips.ethereum.org/EIPS/eip-1#eip-editor-responsibilities

Updated

**Comment by wwhchung** [comment]
@Pandapip1 @lightclient would you be able to take a look at this EIP?  It's very similar to EIP-5131, except as a standalone implementation.

It was built in collaboration with quite a few community members, and has been deployed to:
0x00000000b1BBFe1BF5C5934c4bb9c30FEF15E57A

On mainnet and goerli.

Frontend prototype can be found here:
https://token-delegation.netlify.app/

File: assets/eip-5639/DelegationRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,449 @@
+// SPDX-License-Identifier: MIT
```
Comment (Pandapip1) [inline assets/eip-5639/DelegationRegistry.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: assets/eip-5639/IDelegationRegistry.sol [hunk 2]
```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: MIT
```
Comment (Pandapip1) [inline assets/eip-5639/IDelegationRegistry.sol:1]:
```suggestion
// SPDX-License-Identifier: CC0-1.0
```

File: assets/eip-5639/DelegationRegistry.sol [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+    {
+        bytes32 delegateHash = keccak256(
+            abi.encode(delegate, vault, contract_, tokenId, vaultVersion[vault], delegateVersion[vault][delegate])
+        );
+        return
+            delegations[vault][vaultVersion[vault]].contains(delegateHash)
+            ? true
+            : checkDelegateForContract(delegate, vault, contract_);
+    }
+}
```
Comment (Pandapip1) [inline assets/eip-5639/DelegationRegistry.sol:449]:
```suggestion
}

```

File: assets/eip-5639/IDelegationRegistry.sol [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+     * @param delegate The hotwallet to act on your behalf
+     * @param contract_ The address for the contract you're delegating
+     * @param tokenId The token id for the token you're delegating
+     * @param vault The cold wallet who issued the delegation
+     */
+    function checkDelegateForToken(address delegate, address vault, address contract_, uint256 tokenId)
+        external
+        view
+        returns (bool);
+}
```
Comment (Pandapip1) [inline assets/eip-5639/IDelegationRegistry.sol:184]:
```suggestion
}

```

File: EIPS/eip-5639.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+ * purplehat (artblocks)
+ * ryley-o (artblocks)
+ * andy8052 (tessera)
+ * punk6529 (open metaverse)
+ * loopify (loopiverse)
+ * emiliano (nftrentals)
+ * arran (proof)
+ * james (collabland)
+ * john (gnosis safe)
+ * 0xrusowsky
```
Comment (Pandapip1) [inline EIPS/eip-5639.md:289]:
```suggestion
```

File: EIPS/eip-5639.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+
+## Reference Implementation
+0xfoobar token-delegation - a reference implementation
+- MIT licensed, so you can freely use it for your projects
+- Includes test cases
```
Comment (Pandapip1) [inline EIPS/eip-5639.md:294]:
```suggestion
```
Must link to the solidity files. Test cases must be in a separate section.

File: EIPS/eip-5639.md [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+     * @param vault The cold wallet who issued the delegation
+     */
+    function checkDelegateForToken(address delegate, address vault, address contract_, uint256 tokenId)
+        external
+        view
+        returns (bool);
+}
+```
+
+### Checking Delegation ###
```
Comment (Pandapip1) [inline EIPS/eip-5639.md:253]:
```suggestion
### Checking Delegation
```

File: EIPS/eip-5639.md [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.
+
+Let:
+ - `vault` represent the vault address we are trying to authenticate or prove asset ownership for.
+ - `delegate` represent the address we want to use for signing in lieu of `vault`.
+
+
+**A Delegation Registry must implement IDelegationRegistry**
+
+```
```
Comment (Pandapip1) [inline EIPS/eip-5639.md:69]:
```suggestion
```solidity
```

**Review by Pandapip1 — CHANGES_REQUESTED** [review]

**Comment by wwhchung** [comment]
Accepted all changes @Pandapip1  

thanks! Any thoughts on getting this merged in?

**Review by Pandapip1 — APPROVED** [review]

**Pull request merged**